File: lib/clusterbuster/pod_files/sysbench.pl
#!/usr/bin/perl

use POSIX;
use strict;
use File::Path qw(make_path);
my ($dir) = $ENV{'BAK_CONFIGMAP'};
require "$dir/clientlib.pl";

my ($processes, $rundir, $runtime, $sysbench_generic_args, $sysbench_cmd, $sysbench_fileio_args, $sysbench_modes) = parse_command_line(@ARGV);

$SIG{TERM} = sub() { docleanup() };

initialize_timing();
my ($localrundir) = sprintf('%s/%s/%d', $rundir, podname(), $$);

sub removeRundir() {
    if (-d "$localrundir") {
	open(CLEANUP, "-|", "rm -rf '$localrundir'");
	while (<CLEANUP>) {
	    1;
	}
	close(CLEANUP);
    }
}

sub docleanup()  {
    print STDERR "CLEANUP\n";
    removeRundir();
    kill 'KILL', -1;
    POSIX::_exit(0);
}

my (%units_multiplier) = (
    'kb'  => 1000,
    'kib' => 1024,
    'mb'  => 1000 * 1000,
    'mib' => 1024 * 1024,
    'gb'  => 1000 * 1000 * 1000,
    'gib' => 1024 * 1024 * 1024,
    'tb'  => 1000 * 1000 * 1000 * 1000,
    'tib' => 1024 * 1024 * 1024 * 1024,
    );

sub runit() {
    my ($files) = 0;
    my ($totalbytes) = 0;
    my ($seconds) = 0;
    my ($rate) = 0;
    my ($et) = 0;
    my (@known_sysbench_fileio_modes) = qw(seqwr seqrewr seqrd rndrd rndwr rndrw);
    my ($iterations) = 0;
    my ($loops_per_iteration) = 10000;
    if ($sysbench_modes) {
        @known_sysbench_fileio_modes = split(/ +/, $sysbench_modes);
    }
    timestamp(join("|", @known_sysbench_fileio_modes));
    removeRundir();
    if (! make_path($localrundir)) {
        timestamp("Cannot create run directory $localrundir: $!");
	exit(1);
    }
    if (! chdir($localrundir)) {
        timestamp("Cannot cd $localrundir: $!");
	exit(1);
    }
    my (%op_answers) = ();
    my ($firsttime) = 1;
    my ($avgcpu) = 0;
    my ($weight) = .25;
    my ($icputime);
    my ($interval) = 5;
    my $data_start_time = xtime();

    my ($base0_user, $base0_sys) = cputime();
    foreach my $mode (@known_sysbench_fileio_modes) {
	do_sync(idname("$mode+prepare"));
	timestamp("Preparing...");
	timestamp("sysbench --time=$runtime $sysbench_generic_args $sysbench_cmd prepare --file-test-mode=$mode $sysbench_fileio_args");
	open(PREPARE, "-|", "sysbench --time=$runtime $sysbench_generic_args $sysbench_cmd prepare --file-test-mode=$mode $sysbench_fileio_args") || die "Can't run sysbench: $!\n";
	while (<PREPARE>) {
	    if ($_ =~ /^([[:digit:]]+) bytes written in ([[:digit:].]+) seconds/) {
		$totalbytes = $1;
		$seconds = $2;
		$rate = $1 / $2;
	    } elsif ($_ =~ /^Creating file /) {
		$files++;
		next;
	    }
	    if ($ENV{"VERBOSE"} > 0) {
		chomp;
		timestamp($_);
	    }
	}
	close PREPARE;

	do_sync(idname("$mode+start"));
	my ($op0_user, $op0_sys) = cputime();
	timestamp("Running...");
	timestamp("sysbench --time=$runtime $sysbench_generic_args $sysbench_cmd run --file-test-mode=$mode $sysbench_fileio_args");
	open(RUN, "-|", "sysbench --time=$runtime $sysbench_generic_args $sysbench_cmd run --file-test-mode=$mode $sysbench_fileio_args") || die "Can't run sysbench: $!\n";
	my (%op_answer) = (
	    'final_fsync_enabled' => 'Disabled',
	    'io_mode' => 'unknown',
	    'rdwr_ratio' => 1.0,
	    );
	while (<RUN>) {
	    if      ($_ =~ m;^[[:space:]]*([[:digit:]]+) *files, *([[:digit:]]+)([KMGT]i?B);) {
		$op_answer{'files'} = $1 + 0.0;
		$op_answer{'filesize'} = $2 * (defined $units_multiplier{lc $3} ? $units_multiplier{lc $3} : 1);
	    } elsif ($_ =~ m;^[[:space:]]*Block size *([[:digit:]]+)([KMGT]i?B);) {
		$op_answer{'blocksize'} = $1 * (defined $units_multiplier{lc $2} ? $units_multiplier{lc $2} : 1);
	    } elsif ($_ =~ m;^[[:space:]]*Read/Write ratio for combined random IO test: *([[:digit:]]+(\.[[:digit:]]+)?);) {
		$op_answer{'rdwr_ratio'} = $1 * 1.0;
	    } elsif ($_ =~ m;^[[:space:]]*Periodic FSYNC enabled, calling fsync\(\) each ([[:digit:]]+);) {
		$op_answer{'fsync_frequency'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*calling fsync\(\) at the end of test, (enabled|disabled);i) {
		$op_answer{'final_fsync_enabled'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*Using (.*) I/O mode;) {
		$op_answer{'io_mode'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*reads/s:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'read_ops'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*writes/s:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'write_ops'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*fsyncs/s:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'fsync_ops'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*read, MiB/s:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'read_rate_mb_sec'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*written, MiB/s:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'write_rate_mb_sec'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*total time:[[:space:]]*([[:digit:].]+)s;) {
		$op_answer{'elapsed_time'} = $1 + 0.0;
	    } elsif ($_ =~ m;^[[:space:]]*min:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'min_latency_sec'} = $1 / 1000.0;
	    } elsif ($_ =~ m;^[[:space:]]*avg:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'avg_latency_sec'} = $1 / 1000.0;
	    } elsif ($_ =~ m;^[[:space:]]*max:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'max_latency_sec'} = $1 / 1000.0;
	    } elsif ($_ =~ m;^[[:space:]]*95th percentile:[[:space:]]*([[:digit:].]+);) {
		$op_answer{'p95_latency_sec'} = $1 / 1000.0;
	    }
	    if ($ENV{"VERBOSE"} > 0) {
		chomp;
		timestamp($_);
	    }
	}
	close RUN;
	wait();
	my ($op1_user, $op1_sys) = cputime();
	$op_answer{'user_cpu_time'} = $op1_user - $op0_user;
	$op_answer{'sys_cpu_time'} = $op1_sys - $op0_sys;
	do_sync(idname("$mode+finish"));
	timestamp("Cleanup...");
	timestamp("sysbench --time=$runtime $sysbench_generic_args $sysbench_cmd cleanup --file-test-mode=$mode $sysbench_fileio_args");
	open(CLEANUP, "-|", "sysbench --time=$runtime $sysbench_generic_args $sysbench_cmd cleanup --file-test-mode=$mode $sysbench_fileio_args") || die "Can't run sysbench: $!\n";
	while (<CLEANUP>) {
	    if ($ENV{"VERBOSE"} > 0) {
		chomp;
		timestamp($_);
	    }
	}
	close CLEANUP;
	$op_answers{$mode} = \%op_answer;
    }
    my $data_end_time = xtime();
    my ($elapsed_time) = $data_end_time - $data_start_time;
    my ($base1_user, $base1_sys) = cputime();
    my ($user) = $base1_user - $base0_user;
    my ($sys) = $base1_sys - $base0_sys;
    my (%extras) = (
	'workloads' => \%op_answers
	);
    my ($answer) = print_json_report($data_start_time, $data_end_time, $elapsed_time, $user, $sys, \%extras);
    print STDERR "$answer\n";
    do_sync($answer);
}

run_workload($processes, \&runit);

File: lib/clusterbuster/pod_files/sysbench.py
#!/usr/bin/env python3

import os
import subprocess
import re
import shutil
import time

from clusterbuster_pod_client import clusterbuster_pod_client


class sysbench_client(clusterbuster_pod_client):
    """
    sysbench test for clusterbuster
    """

    def __init__(self):
        try:
            super().__init__()
            self.np = '([0-9]+([kmgt]i)?)b'
            self._set_processes(int(self._args[0]))
            self.rundir = self._args[1]
            self.runtime = int(self._args[2])
            self.workload = self._args[3]
            if self._args[4]:
                self.sysbench_fileio_tests = self._splitStr(r'\s+', self._args[4])
            else:
                self.sysbench_fileio_tests = ['seqwr', 'seqrewr', 'seqrd', 'rndrd', 'rndwr', 'rndrw']
            if self._args[5]:
                self.sysbench_fileio_modes = self._splitStr(r'\s+', self._args[5])
            else:
                self.sysbench_fileio_modes = ['sync']
            self.sysbench_options = self._args[6:]
            self.runit_cpu = self.runit_simple
            self.runit_memory = self.runit_simple
            self.runit_mutex = self.runit_simple
            self.runit_threads = self.runit_simple
        except Exception as err:
            self._abort(f"Init failed! {err} {' '.join(self._args)}")

    def build_sysbench_cmd(self, command: str, *argv):
        args = ['sysbench', f'--time={self.runtime}', self.workload, command]
        args.extend(argv)
        args.extend(self.sysbench_options)
        return args

    def simple_check1(self, m, op_answer: dict, key: str, multiplier: float, matchidx: int,
                      is_float: bool, is_str: bool, is_int: bool):
        if is_str:
            op_answer[key] = m.group(matchidx)
        elif is_float:
            op_answer[key] = float(m.group(matchidx)) * multiplier
        else:
            op_answer[key] = self._toSize(m.group(matchidx)) * multiplier

    def simple_check(self, pattern: str, line: str, op_answer: dict, key, multiplier=None, matchidx=None,
                     is_float: bool = False, is_str: bool = False, is_int: bool = True):
        m = re.match(pattern, line)
        if not m:
            return False
        if isinstance(key, list):
            if multiplier is None:
                multiplier = [1.0 for i in range(len(key))]
            elif isinstance(multiplier, int) or isinstance(multiplier, float):
                multiplier = [multiplier for i in range(len(key))]
            elif not isinstance(matchidx, list) or len(matchidx) < len(key):
                raise Exception("Non-conformant multiplier")
            if not isinstance(matchidx, list) or len(matchidx) < len(key):
                raise Exception("Non-conformant multiplier")
            for i in range(len(key)):
                self.simple_check1(m, op_answer, key[i], multiplier[i], matchidx[i], is_float, is_str, is_int)
        else:
            if matchidx is None:
                matchidx = 1
            if multiplier is None:
                multiplier = 1
            self.simple_check1(m, op_answer, key, multiplier, matchidx, is_float, is_str, is_int)
        return True

    def runit_simple(self, process: int):
        args = self.build_sysbench_cmd('run')
        self._sync_to_controller()
        self._timestamp(f'Running {" ".join(args)}')
        data_start_time = self._adjusted_time()
        user, sys = self._cputimes()
        op_answer = {
            'op_start': self._adjusted_time(),
            'workload': self.workload
            }
        with subprocess.Popen(args, stdout=subprocess.PIPE) as run:
            line = run.stdout.readline().decode('ascii')
            while line:
                line = line.strip().lower()
                self._timestamp(line)
                if self.simple_check(r'number of threads: *([0-9]+)', line, op_answer, 'threads'):
                    pass
                elif self.simple_check(r'events per second: *([0-9.]+)', line, op_answer, 'events_per_second', is_float=True):
                    pass
                elif self.simple_check(r'total time: *([0-9.]+)s', line, op_answer, 'elapsed_time', is_float=True):
                    pass
                elif self.simple_check(r'total number of events: *([0-9]+)', line, op_answer, 'total_events'):
                    pass
                elif self.simple_check(r'min: *([0-9.]+)', line, op_answer, 'min_latency_sec', is_float=True):
                    pass
                elif self.simple_check(r'avg: *([0-9.]+)', line, op_answer, 'avg_latency_sec', is_float=True):
                    pass
                elif self.simple_check(r'max: *([0-9.]+)', line, op_answer, 'max_latency_sec', is_float=True):
                    pass
                elif self.simple_check(r'95th percentile: *([0-9.]+)', line, op_answer, 'p95_latency_sec', is_float=True):
                    pass
                elif self.simple_check(r'events .avg/stddev.:\s*([0-9.]+)/([0-9.]+)', line,
                                       op_answer, ['events_avg', 'events_stdev'],
                                       matchidx=[1, 2], is_float=True):
                    pass
                elif self.simple_check(r'execution time .avg/stddev.:\s*([0-9.]+)/([0-9.]+)', line,
                                       op_answer, ['time', 'time_stdev'],
                                       matchidx=[1, 2], is_float=True):
                    pass
                line = run.stdout.readline().decode('ascii')
            status = run.wait()
            if status:
                raise Exception(f"Sysbench failed: {status}")
        op_answer['op_end'] = self._adjusted_time()
        data_end_time = self._adjusted_time()
        elapsed_time = data_end_time - data_start_time
        user, sys = self._cputimes(user, sys)
        op_answer['user_cpu_time'] = user
        op_answer['sys_cpu_time'] = sys
        extras = {
            'workloads': {self.workload: op_answer}
            }
        self._report_results(data_start_time, data_end_time, elapsed_time, user, sys, extras)

    def runit_fileio(self, process: int):
        op_answers = {}
        localrundir = f'{self.rundir}/{self._podname()}/{str(os.getpid())}'
        shutil.rmtree(localrundir, ignore_errors=True)
        while True:
            try:
                os.makedirs(localrundir)
                break
            except Exception as exc:
                self._timestamp(f"makedirs failed {exc}; retrying")
                time.sleep(1)
        os.chdir(localrundir)
        data_start_time = self._adjusted_time()
        user, sys = self._cputimes()
        for mode in self.sysbench_fileio_modes:
            for test in self.sysbench_fileio_tests:
                self._sync_to_controller(f'{test}+{mode}+prepare')
                args = self.build_sysbench_cmd('prepare', f'--file-test-mode={test}', f'--file-io-mode={mode}')
                self._timestamp(f'Preparing {" ".join(args)}')
                subprocess.run(args, check=True)

                self._drop_cache()
                self._sync_to_controller(f'{test}+{mode}+run')
                op_user, op_sys = self._cputimes()
                args = self.build_sysbench_cmd('run', f'--file-test-mode={test}', f'--file-io-mode={mode}')
                self._timestamp(f'Running {" ".join(args)}')
                op_answer = {
                    'final_fsync_enabled': 'Disabled',
                    'io_mode': 'unknown',
                    'workload': test,
                    'rdwr_ratio': 1,
                    'op_start': self._adjusted_time()
                    }
                with subprocess.Popen(args, stdout=subprocess.PIPE) as run:
                    line = run.stdout.readline().decode('ascii')
                    while line:
                        line = line.strip().lower()
                        self._timestamp(line)
                        if self.simple_check(rf'([0-9]+) *files, *{self.np}', line,
                                             op_answer, ['files', 'filesize'], matchidx=[1, 2]):
                            pass
                        elif self.simple_check(rf'block size *{self.np}', line, op_answer, 'blocksize'):
                            pass
                        elif self.simple_check(r'read/write ratio for combined random io test: *([0-9]+(\.[0-9]+)?)',
                                               line, op_answer, 'rdwr_ratio', is_float=True):
                            pass
                        elif self.simple_check(r'periodic fsync enabled, calling fsync.. each ([0-9]+)',
                                               line, op_answer, 'fsync_frequency'):
                            pass
                        elif self.simple_check(r'calling fsync.. at the end of test, (enabled|disabled)',
                                               line, op_answer, 'final_fsync_enabled', is_str=True):
                            pass
                        elif self.simple_check(r'using (.*) i/o mode', line, op_answer, 'io_mode', is_str=True):
                            pass
                        elif self.simple_check(r'reads/s: *([0-9.]+)', line, op_answer, 'read_ops'):
                            pass
                        elif self.simple_check(r'writes/s: *([0-9.]+)', line, op_answer, 'write_ops'):
                            pass
                        elif self.simple_check(r'fsyncs/s: *([0-9.]+)', line, op_answer, 'fsync_ops'):
                            pass
                        elif self.simple_check(r'read, mib/s: *([0-9.]+)', line, op_answer, 'read_rate_mb_sec'):
                            pass
                        elif self.simple_check(r'written, mib/s: *([0-9.]+)', line, op_answer, 'write_rate_mb_sec'):
                            pass
                        elif self.simple_check(r'total time: *([0-9.]+)s', line, op_answer, 'elapsed_time'):
                            pass
                        elif self.simple_check(r'min: *([0-9.]+)', line, op_answer, 'min_latency_sec'):
                            pass
                        elif self.simple_check(r'avg: *([0-9.]+)', line, op_answer, 'avg_latency_sec'):
                            pass
                        elif self.simple_check(r'max: *([0-9.]+)', line, op_answer, 'max_latency_sec'):
                            pass
                        elif self.simple_check(r'95th percentile: *([0-9.]+)', line, op_answer, 'p95_latency_sec'):
                            pass
                        line = run.stdout.readline().decode('ascii')
                    status = run.wait()
                    if status:
                        raise Exception(f"Sysbench failed: {status}")
                op_answer['op_end'] = self._adjusted_time()
                op_answer['user_cpu_time'], op_answer['sys_cpu_time'] = self._cputimes(op_user, op_sys)
                self._sync_to_controller(f'{test}+{mode}+finish')
                args = self.build_sysbench_cmd('cleanup', f'--file-test-mode={test}', f'--file-io-mode={mode}')
                self._timestamp(f'Cleanup {" ".join(args)}')
                subprocess.run(args, check=True)
                op_answers[f'fileio+{test}+{mode}'] = op_answer
        data_end_time = self._adjusted_time()
        elapsed_time = data_end_time - data_start_time
        user, sys = self._cputimes(user, sys)
        extras = {
            'workloads': op_answers
            }
        self._report_results(data_start_time, data_end_time, elapsed_time, user, sys, extras)

    def runit(self, process: int):
        func = getattr(self, f'runit_{self.workload}', None)
        if func:
            func(process)
        else:
            self.runit_simple(process)


sysbench_client().run_workload()
